DisjointGridSet.java

```diff
@@ -0,0 +1,57 @@
+package assignment4;
+
+import java.util.HashSet;
+import java.util.Set;
+
+
+public class DisjointGridSet {
+
+    private int[] parents;
+    private int initialColumnCount;
+
+    public DisjointGridSet(boolean[][] grid) {
+        this.initialColumnCount = grid[0].length;
+        this.parents = new int[grid.length * grid[0].length];
+        for (int i = 0; i < grid.length; i++) {
+            for (int j = 0; j < grid[0].length; j++) {
+                // initially every island cell is marked as a separate island
+                // -1 for water
+                this.parents[convertIndexes(i, j)] = grid[i][j] ? convertIndexes(i, j) : -1;
+            }
+        }
+    }
+
+    public int find(int element) {
+        if (element == parents[element]) {
+            return element;
+        } else {
+            parents[element] = find(parents[element]);
+            return parents[element];
+        }
+    }
+
+    public void union(int element1, int element2) {
+        element1 = find(element1);
+        element2 = find(element2);
+        if (element1 == element2) {
+            return;
+        }
+        parents[element1] = element2;

```

Comment:
This code makes me a little nervous as a reader. The nerves come from "how do I know we do not get a circle where parent{1] = 2. parent[2] = 3, parent[3] = 1, and then we would have an infinite loop in find.

If you added comparison logic so that parent[element1] < element1 (or the other way around), then it would guarantee you never get an infinite loop, and so reasoning about this would be easier.

I think this works ok because of the order you union, but it isn't obviously correct, and _obviously_ correct is much better than just being right.